File: Tree/173.Binary_Search_Tree_Iterator/Binary_Search_Tree_Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    int ptr = 0;
    vector<int> Nodes = {-1};
    
    BSTIterator(TreeNode* root) {
        INORDER_DFS(root);
    }
    
    void INORDER_DFS(TreeNode* root){
        if(!root) return;
        INORDER_DFS(root->left);
        Nodes.push_back(root->val);
        INORDER_DFS(root->right);
    }

    int next() {
        ptr++;
        return Nodes[ptr];
    }
    
    bool hasNext() {
        return ptr+1 < Nodes.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Tree/173.Binary_Search_Tree_Iterator/Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.ptr = 0
        self.Nodes = [-1]
        self.DFS(root)
        self.Nodes_length = len(self.Nodes)
    def DFS(self, root):
        if(root == None): return 
        self.DFS(root.left)
        self.Nodes.append(root.val)
        self.DFS(root.right)
    def next(self) -> int:
        self.ptr += 1
        return self.Nodes[self.ptr]

    def hasNext(self) -> bool:
        return self.ptr + 1 < self.Nodes_length 


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()